std/range/package.d

```diff
@@ -8606,6 +8606,385 @@ if (isForwardRange!Source && hasLength!Source)
+            }
+
+            static if (isInfinite!Source && hasLength!Indexes)
+            /// Length is only provided for infinite sources, when hasLength!Indexes is true
+                @property size_t length()
+                {
+                    return _endIndexes.length + 1;
+                }
+
+            static if (hasSlicing!Source && hasSlicing!Indexes && hasLength!Indexes)
+            {
+                /**
+                Indexing and slicing operations are only provided when both Source and Indexes
+                have slicing and hasLength!Indexes is true.
+                */
+                auto opIndex(size_t index)
+                {
+                    assert(index<=_endIndexes.length, "Chunks index out of bounds");
+
+                    const size_t startIndex = (index == 0 ? 0 : _endIndexes[index-1] - _offset);
+
+                    static if (hasLength!Source)
+                        assert(startIndex<=_source.length, "Chunks index out of bounds");
+
+                    if (index == _endIndexes.length)
+                        return _source[startIndex .. $];
+
+                    static if (hasLength!Source)
+                    {
+                        import std.algorithm.comparison : min;
+                        const size_t endIndex = min(_endIndexes[index] - _offset, source.length);
+                    }
+                    else
+                        const size_t endIndex = _endIndexes[index] - _offset;
+
+                    return _source[startIndex .. endIndex];
+                }
+
+                /// Ditto
+                auto opSlice(size_t lower, size_t upper)
+                {
+                    assert(lower <= upper, "Start index must be lower than end index");
+                    assert(lower <= _endIndexes.length+1 && upper <= _endIndexes.length+1,
+                        "Chunks index out of bounds");
+
+                    if (lower == upper || lower == _endIndexes.length+1)
+                        return Chunks(_source[$ .. $], _endIndexes[$ .. $]);
+
+                    const size_t startIndex = (lower == 0 ? 0 : _endIndexes[lower-1] - _offset);
+
+                    if (upper >= _endIndexes.length)
+                        return Chunks(_source[startIndex .. $], _endIndexes[lower .. $], startIndex,
+                            UseMaxLength(true, upper-lower));
+
+                    static if (hasLength!Source)
+                    {
+                        import std.algorithm.comparison : min;
+                        const size_t endIndex = min(_endIndexes[upper-1] - _offset, source.length);
+                    }
+                    else
+                        const size_t endIndex = _endIndexes[upper-1] - _offset;
+
+                    return Chunks(_source[startIndex .. endIndex], _endIndexes[lower .. upper-1], startIndex);
+                }
+            }
+
+            // Used when slicing to avoid taking extra elements
```

Comment:
I'm not understanding why this is needed. Can you give an example?

Reply:
I was thinking of this case: when the source range is longer than the last index provided and the requested slice should end just before the chunk containing the extra elements:
source = [1, 2, 3, 4, 5]
indexes = [1, 2, 3]
chunks[1 .. 3] should be [[2], [3]]
But since the whole range is supposed to also include one last chunk with the extra items, when returning the new range without specifying the needed length chunks[1 .. 3] would turn into [[2], [3], [4, 5]]. This should only be a problem when hasLength!Source and isInfinite!Source are both false since in this case we can't check if the end index is still in range. Actually it might be safer to only allow indexing in these two cases.

Reply:
Ah, ok, I get it. I think all you need is a boolean saying whether the tail should be returned.

Reply:
Please include this example in the unittests (if not already).